File: 1-to-1 Video/Agora-macOS-Tutorial-Objective-C-1to1/Agora Mac Tutorial Objective-C/VideoChat/DeviceSelectionViewController.m
//
//  DeviceSelectionViewController.m
//  Agora Mac Tutorial Objective-C
//
//  Created by James Fang on 8/9/16.
//  Copyright © 2016 Agora.io. All rights reserved.
//

#import "DeviceSelectionViewController.h"

@interface DeviceSelectionViewController ()

@property (weak) IBOutlet NSPopUpButton *microphoneSelection;
@property (weak) IBOutlet NSPopUpButton *speakerSelection;
@property (weak) IBOutlet NSPopUpButton *cameraSelection;

@property (strong, nonatomic) NSArray* connectedRecordingDevices;
@property (strong, nonatomic) NSArray* connectedPlaybackDevices;
@property (strong, nonatomic) NSArray* connectedVideoCaptureDevices;

@end

@implementation DeviceSelectionViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    // Tutorial Step 11
    self.preferredContentSize = NSMakeSize(500, 250);
    [self loadDevicesInPopUpButtons];
}

- (void)loadDevicesInPopUpButtons {
    [self.microphoneSelection removeAllItems];
    [self.speakerSelection removeAllItems];
    [self.cameraSelection removeAllItems];

    self.connectedRecordingDevices = [self.agoraKit enumerateDevices:0];
    for (id device in self.connectedRecordingDevices) {
        [self.microphoneSelection addItemWithTitle:([device deviceName])];
    }
    
    self.connectedPlaybackDevices = [self.agoraKit enumerateDevices:1];
    for (id device in self.connectedPlaybackDevices) {
        [self.speakerSelection addItemWithTitle:([device deviceName])];
    }
    
    self.connectedVideoCaptureDevices = [self.agoraKit enumerateDevices:3];
    for (id device in self.connectedVideoCaptureDevices) {
        [self.cameraSelection addItemWithTitle:([device deviceName])];
    // Populate the NSPopUpButtons with the enumerated device list
    }
}

- (void)rtcEngine:(AgoraRtcEngineKit *)engine device:(NSString*) deviceId type:(AgoraMediaDeviceType) deviceType stateChanged:(NSInteger) state {
    [self loadDevicesInPopUpButtons];
    // Repopulate NSPopUpButtons if a device is plugged in or removed
}

- (IBAction)didClickConfirmButton:(NSButton *)button {
    [self.agoraKit setDevice:0 deviceId:[self.connectedRecordingDevices[self.microphoneSelection.indexOfSelectedItem] deviceId]];
    [self.agoraKit setDevice:1 deviceId:[self.connectedPlaybackDevices[self.speakerSelection.indexOfSelectedItem] deviceId]];
    [self.agoraKit setDevice:3 deviceId:[self.connectedVideoCaptureDevices[self.cameraSelection.indexOfSelectedItem] deviceId]];
    // Set the devices
    [self dismissViewController:self];
}

@end

File: 1-to-1 Video/Agora-macOS-Tutorial-Swift-1to1/Agora Mac Tutorial Swift/VideoChat/DeviceSelectionViewController.swift
//
//  DeviceSelectionViewController.swift
//  Agora Mac Tutorial Swift
//
//  Created by James Fang on 8/12/16.
//  Copyright © 2016 Agora.io. All rights reserved.
//

import Cocoa

protocol DeviceSelectionDelegate: NSObjectProtocol {
}

class DeviceSelectionViewController: NSViewController {
    @IBOutlet weak var microphoneSelection: NSPopUpButton!
    @IBOutlet weak var speakerSelection: NSPopUpButton!
    @IBOutlet weak var cameraSelection: NSPopUpButton!
    
    var connectedRecordingDevices: NSArray!
    var connectedPlaybackDevices: NSArray!
    var connectedVideoCaptureDevices: NSArray!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.preferredContentSize = NSMakeSize(500, 250)
        loadDevicesInPopUpButtons()
    }
    
    func loadDevicesInPopUpButtons() {
        microphoneSelection.removeAllItems()
        speakerSelection.removeAllItems()
        cameraSelection.removeAllItems()
    
        connectedRecordingDevices = AgoraKit.enumerateDevices(.audioRecording) as NSArray!
        for device in connectedRecordingDevices {
            if let title = (device as AnyObject).deviceName as? String {
                microphoneSelection.addItem(withTitle: title)
            }
        }
    
        connectedPlaybackDevices = AgoraKit.enumerateDevices(.audioPlayout) as NSArray!
        for device in connectedPlaybackDevices {
            if let title = (device as AnyObject).deviceName as? String {
                speakerSelection.addItem(withTitle: title)
            }
        }
        
        connectedVideoCaptureDevices = AgoraKit.enumerateDevices(.videoCapture) as NSArray!
        for device in connectedVideoCaptureDevices {
            if let title = (device as AnyObject).deviceName as? String {
                cameraSelection.addItem(withTitle: title)
            }
        }
        // Populate the NSPopUpButtons with the enumerated device list
    }
    
    @IBAction func didClickConfirmButton(_ sender: NSButton) {
        if let deviceId = (connectedRecordingDevices[microphoneSelection.indexOfSelectedItem] as AnyObject).deviceId as? String {
            AgoraKit.setDevice(.audioRecording, deviceId: deviceId)
        }
        if let deviceId = (connectedPlaybackDevices[speakerSelection.indexOfSelectedItem] as AnyObject).deviceId as? String {
            AgoraKit.setDevice(.audioPlayout, deviceId: deviceId)
        }
        if let deviceId = (connectedVideoCaptureDevices[cameraSelection.indexOfSelectedItem] as AnyObject).deviceId as? String {
            AgoraKit.setDevice(.videoCapture, deviceId: deviceId)
        }
        // Set the devices
        self.dismiss(self)
    }

}

extension DeviceSelectionViewController: AgoraRtcEngineDelegate {
    @objc(rtcEngine:device:type:stateChanged:) func rtcEngine(_ engine: AgoraRtcEngineKit, device deviceId: String, type deviceType: AgoraMediaDeviceType, stateChanged state: Int) {
        loadDevicesInPopUpButtons()
        // Repopulate NSPopUpButtons if a device is plugged in or removed
    }
}
